plugins/outputs/syslog/syslog_mapper.go

```diff
@@ -0,0 +1,193 @@
+		return nil, errors.New("metric could not produce valid syslog message")
+	}
+	return msg, nil
+}
+
+func (sm *SyslogMapper) mapStructuredData(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	for key, value := range metric.Fields() {
+		if sm.reservedFields[key] {
+			continue
+		}
+		isExplicitSdid := false
+		for _, sdid := range sm.Sdids {
+			k := strings.TrimLeft(key, sdid+sm.Separator)
+			if len(key) > len(k) {
+				isExplicitSdid = true
+				msg.SetParameter(sdid, k, formatValue(value))
+				break
+			}
+		}
+		if !isExplicitSdid && len(sm.DefaultSdid) > 0 {
+			k := strings.TrimLeft(key, sm.DefaultSdid+sm.Separator)
+			msg.SetParameter(sm.DefaultSdid, k, formatValue(value))
+		}
+	}
+}
+
+func (sm *SyslogMapper) mapAppname(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	if value, ok := metric.GetTag("appname"); ok {
+		msg.SetAppname(formatValue(value))
+	} else {
+		//Use default appname
+		msg.SetAppname(sm.DefaultAppname)
+	}
+}
+
+func mapMsgID(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	if value, ok := metric.GetField("msgid"); ok {
+		msg.SetMsgID(formatValue(value))
+	} else {
+		// We default to metric name
+		msg.SetMsgID(metric.Name())
+	}
+}
+
+func mapVersion(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	if value, ok := metric.GetField("version"); ok {
+		if v, err := parseIntValue(value); err == nil {
+			msg.SetVersion(uint16(v))
+			return
+		}
+	}
+	msg.SetVersion(1)
+}
+
+func mapMsg(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	if value, ok := metric.GetField("msg"); ok {
+		msg.SetMessage(formatValue(value))
+	}
+}
+
+func mapProcID(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	if value, ok := metric.GetField("procid"); ok {
+		msg.SetProcID(formatValue(value))
+	}
+}
+
+func (sm *SyslogMapper) mapPriority(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	severityCode := sm.DefaultSeverityCode
+	facilityCode := sm.DefaultFacilityCode
+
+	if value, ok := getFieldCode(metric, "severity_code"); ok {
+		severityCode = *value
+	}
+
+	if value, ok := getFieldCode(metric, "facility_code"); ok {
+		facilityCode = *value
+	}
+
+	priority := (8 * facilityCode) + severityCode
+	msg.SetPriority(priority)
+}
+
+func mapHostname(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	// Try with hostname, then with source, then with host tags, then take OS Hostname
+	if value, ok := metric.GetTag("hostname"); ok {
+		msg.SetHostname(formatValue(value))
+	} else if value, ok := metric.GetTag("source"); ok {
+		msg.SetHostname(formatValue(value))
+	} else if value, ok := metric.GetTag("host"); ok {
+		msg.SetHostname(formatValue(value))
+	} else if value, err := os.Hostname(); err == nil {
+		msg.SetHostname(value)
+	}
+}
+
+func mapTimestamp(metric telegraf.Metric, msg *rfc5424.SyslogMessage) {
+	timestamp := metric.Time()
+	if value, ok := metric.GetField("timestamp"); ok {
+		if v, err := parseIntValue(value); err == nil {
+			timestamp = time.Unix(v, 0).UTC()
```

Comment:
This field should already be an int64, if not I think we should skip over it.  Since it is in Nanoseconds you will need to use `time.Unix(0, v)`, the UTC() call is harmless but unnecessary since it is already in UTC.